### PR TITLE
👌 IMPROVE: Use `sqlalchemy.func` for JSONB QB filters

### PR DESCRIPTION
**Pull request by ltalirz**
related to #5282 

The QueryBuilder declared a custom `jsonb_typeof` FunctionElement,
for which sqlalchemy disables query caching by default.

### CODE REVIEW

**Comment by ltalirz** [comment]
Removes the warning locally for me

**Comment by chrisjsewell** [comment]
Cheers, can you try something of the lines of:

```python
QueryBuilder().append(Node, filters={"attributes.x": {'of_type': 'bool'}}).as_sql()`
QueryBuilder().append(Node, filters={"attributes.x": {'of_length': 3}}).as_sql()`
```

just to check that this still compiles correctly.
I seem to remember some issues around this previously

**Comment by ltalirz** [comment]
You mean this? (note: replaced 'bool' => 'boolean')
```ipython
In [2]: QueryBuilder().append(Node, filters={"attributes.x": {'of_type': 'boolean'}}).as_sql()
Out[2]: "'SELECT db_dbnode_1.id, db_dbnode_1.uuid, db_dbnode_1.node_type, db_dbnode_1.process_type, db_dbnode_1.label, db_dbnode_1.description, db_dbnode_1.ctime, db_dbnode_1.mtime, db_dbnode_1.attributes, db_dbnode_1.extras, db_dbnode_1.repository_metadata, db_dbnode_1.dbcomputer_id, db_dbnode_1.user_id \\nFROM db_dbnode AS db_dbnode_1 \\nWHERE CAST(db_dbnode_1.node_type AS VARCHAR) LIKE %(param_1)s AND jsonb_typeof((db_dbnode_1.attributes #> %(attributes_1)s)) = %(jsonb_typeof_1)s' % {'param_1': '%', 'attributes_1': ('x',), 'jsonb_typeof_1': 'boolean'}\n"

In [3]: QueryBuilder().append(Node, filters={"attributes.x": {'of_length': 3}}).as_sql()
Out[3]: "'SELECT db_dbnode_1.id, db_dbnode_1.uuid, db_dbnode_1.node_type, db_dbnode_1.process_type, db_dbnode_1.label, db_dbnode_1.description, db_dbnode_1.ctime, db_dbnode_1.mtime, db_dbnode_1.attributes, db_dbnode_1.extras, db_dbnode_1.repository_metadata, db_dbnode_1.dbcomputer_id, db_dbnode_1.user_id \\nFROM db_dbnode AS db_dbnode_1 \\nWHERE CAST(db_dbnode_1.node_type AS VARCHAR) LIKE %(param_1)s AND CASE WHEN (jsonb_typeof((db_dbnode_1.attributes #> %(attributes_1)s)) = %(jsonb_typeof_1)s) THEN jsonb_array_length(CAST((db_dbnode_1.attributes #> %(attributes_1)s) AS JSONB)) = %(jsonb_array_length_1)s ELSE %(param_2)s END' % {'param_1': '%', 'attributes_1': ('x',), 'jsonb_typeof_1': 'array', 'jsonb_array_length_1': 3, 'param_2': False}\n"
```

The one failing tests is about a change in the SQL representation
```
E       -'SELECT db_dbnode_1.id, db_dbnode_1.uuid \nFROM db_dbnode AS db_dbnode_1 \nWHERE CAST(db_dbnode_1.node_type AS VARCHAR) LIKE %(param_1)s AND CASE WHEN (jsonb_typeof((db_dbnode_1.extras #> %(extras_1)s)) = %(param_2)s) THEN (db_dbnode_1.extras #>> %(extras_1)s) = %(param_3)s ELSE %(param_4)s END' % {'param_1': '%', 'extras_1': ('tag4',), 'param_2': 'string', 'param_3': 'appl_pecoal', 'param_4': False}
E       +'SELECT db_dbnode_1.id, db_dbnode_1.uuid \nFROM db_dbnode AS db_dbnode_1 \nWHERE CAST(db_dbnode_1.node_type AS VARCHAR) LIKE %(param_1)s AND CASE WHEN (jsonb_typeof((db_dbnode_1.extras #> %(extras_1)s)) = %(jsonb_typeof_1)s) THEN (db_dbnode_1.extras #>> %(extras_1)s) = %(param_2)s ELSE %(param_3)s END' % {'param_1': '%', 'extras_1': ('tag4',), 'jsonb_typeof_1': 'string', 'param_2': 'appl_pecoal', 'param_3': False}
``` 
but I think it's just about naming - the placeholder that used to be `param_2` is now called `jsonb_typeof_1`

File: aiida/storage/psql_dos/orm/querybuilder/main.py [hunk 1]
```diff
@@ -367,7 +333,7 @@ def _build(self) -> Query:
             # LINK-PROJECTIONS #########################
 
             for vertex in self._data['path'][1:]:
-                edge_tag = vertex.get('edge_tag', None)  # type: ignore
+                edge_tag = vertex.get('edge_tag', None)
```
Comment (chrisjsewell) [inline aiida/storage/psql_dos/orm/querybuilder/main.py:336]: @ltalirz you changed this, hence the failing mypy

Comment (ltalirz) [reply aiida/storage/psql_dos/orm/querybuilder/main.py:336]: cheers just realized it. I still don't quite get why mypy locally and on ci often gives different results

**Comment by chrisjsewell** [comment]
> Do I need to update some test data to make this pass?

Yeh the one you mentioned already: `tests/orm/test_querybuilder.py::TestRepresentations::test_as_sql`

The change looks fine, its a pytest-regressions test, so you can even run `tox tests/orm/test_querybuilder.py::TestRepresentations::test_as_sql -- --force-regen` to update it

**Comment by ltalirz** [comment]
Great, thanks for the hints!

**Comment by chrisjsewell** [comment]
Maybe just change the PR/commit name to something like ``👌 IMPROVE: Use `sqlalchemy.func` for JSONB QB filters``

**Comment by ltalirz** [comment]
Done! I'll still need your approving review for this to auto-merge

**Comment by chrisjsewell** [comment]
> I'll still need your approving review for this to auto-merge

change the PR name ☝️

**Review by chrisjsewell — APPROVED** [review]
cheers!

**Pull request merged**